app/src/control_msg.c

```diff
@@ -93,6 +140,94 @@ control_msg_serialize(const struct control_msg *msg, unsigned char *buf) {
     }
 }
 
+void
+control_msg_log(const struct control_msg *msg) {
+#define LOG_CMSG(fmt, ...) LOGV("input: " fmt, ## __VA_ARGS__)
+    switch (msg->type) {
+        case CONTROL_MSG_TYPE_INJECT_KEYCODE:
+            LOG_CMSG("key %-4s code=%d repeat=%" PRIu32 " meta=%06lx",
+                     KEYEVENT_ACTION_LABEL(msg->inject_keycode.action),
+                     (int) msg->inject_keycode.keycode,
+                     msg->inject_keycode.repeat,
+                     (long) msg->inject_keycode.metastate);
+            break;
+        case CONTROL_MSG_TYPE_INJECT_TEXT:
+            LOG_CMSG("text \"%s\"", msg->inject_text.text);
+            break;
+        case CONTROL_MSG_TYPE_INJECT_TOUCH_EVENT: {
+            int action = msg->inject_touch_event.action
+                       & AMOTION_EVENT_ACTION_MASK;
+            uint64_t id = msg->inject_touch_event.pointer_id;
+            if (id == POINTER_ID_MOUSE || id == POINTER_ID_VIRTUAL_FINGER) {
+                // string pointer id
+                LOG_CMSG("touch [id=%s] %-4s position=%" PRIi32 ",%" PRIi32
+                             " pressure=%f buttons=%06lx",
```

Comment:
I used `%g` for pressure so it hides the fractional part when there are no digits. I guess you changed this intentionally?

`pressure=0.000000` vs `pressure=0`

Reply:
Oh, yes, I forgot to discuss this change, the reason was that `%g` prints exponential notation when necessary, and I think we don't want that here:

```c
printf("%f %g\n", 0.0000001f, 0.0000001f);
```

prints:

```
0.000000 1e-07
```

But yes, `%g` is probably better in practice. I will change back to `%g`.

I also changed `position=500x400` to `position=500,400`, since `500x400` suggests a dimensions rather than a position.

Reply:
Pushed to `logv`.

Reply:
Agreed, the `,` separator makes sense.